Add missing stm32h7 ethernet

ehntoo opened the pull request:
Addresses Issue #41 

This PR pulls in all the missing ethernet registers from the stm32h7 reference manual (RM0433 https://www.st.com/resource/en/reference_manual/dm00314099.pdf)

Some things to note - this ethernet peripheral's registers are grouped significantly differently from others in the STM32 family. I've reproduced the DMA, MAC, MTL register groups from the manual, but added a separate MMC group, as previous family members seem to have had.

There are also a number of ethernet statistic registers such as "TX_PACKET_COUNT_GOOD" which do not share the "MAC" register prefix - I have left them in the "Ethernet_MAC" register group.

I also noticed an aberration in addressing for the MACARPAR register - it looks from the register map as if it should be at offset 0xB10, but the manual is consistent in placing it at 0xAE0. ST's own HAL code puts it at 0xB10 (https://raw.githubusercontent.com/fhsakaci/STM32H743-LwIP/master/STM32Cube_FW_H7_V1.3.0/Drivers/CMSIS/Device/ST/STM32H7xx/Include/stm32h743xx.h), so I've chalked that up as an error in the reference manual.

Let me know if there are any changes I should make or if you'd like the commits squashed. :)

adamgreig commented:
This looks great, thanks! Did you automatically generate it?

I'm going to check it against the ref man this weekend. It might be worth moving the ethernet stuff into its own file, both since there's a lot of it and since the upcoming STM32H value-line devices might have the same problem. You can see how a few parts (e.g. the stm32f7 already) use the `common_patches` folder and just include YAML from there.

ehntoo commented:
It's semi-generated using copy-and-paste and a whole lot of Vim macro abuse. There's a little room for error, but things should be _fairly_ accurate. :) Were I to do this again, I think I'd spend some more time and put together a proper script to scrape things out of the reference manual in a more automated fashion.

I'll pull things into a separate patch file tomorrow, it seems like a good way to approach this.

ehntoo commented:
Well, "tomorrow" turned into 2 and a half weeks, but I finally got back to this. I extracted the modifications into a separate patch file, which makes the device file _much_ nicer to read again.

I did [report this](https://community.st.com/s/question/0D50X00009hnZLRSA2/stm32h7-svd-bugs) on the ST community forums, and it's now been "reported internally", but we'll see if/when anything comes of that.

adamgreig commented on devices/common_patches/h7_ethernet.yaml:
```diff
@@ -0,0 +1,2082 @@
+        description: Ethernet MAC LPI timers control register
+        addressOffset: 0xD4
+        access: read-write
+        resetValue: 0x00000000
```
This one is `0x03E80000`

adamgreig commented on devices/common_patches/h7_ethernet.yaml:
```diff
@@ -0,0 +1,2082 @@
+        resetValue: 0x00000000
+        access: read-only
+        fields:
+          TSSS:
```
Should be `TSS`

adamgreig commented on devices/common_patches/h7_ethernet.yaml:
```diff
@@ -0,0 +1,2082 @@
+          TSSS:
+            description: Timestamp Sub-seconds
+            bitOffset: 0 
+            bitWidth: 31
```
Should be 32

adamgreig commented on devices/common_patches/h7_ethernet.yaml:
```diff
@@ -0,0 +1,2082 @@
+            description: PPS Output Frequency Control
+            bitOffset: 0 
+            bitWidth: 4
+      MACPPSCR:
```
I don't understand what's going on here and briefly thought I was going mad. The reference manual lists `MACPPSCR` twice, both at the same offset, with the same `TRGTMODSEL0` and `PPSEN0` but totally different `PPSCTRL`/`PPSCMD` in the bottom 4 bits. We can't just list both registers identically with the same offset but who knows what the correct field is. Any ideas?

ehntoo replied:
This had me super confused at first after you pointed it out - it looks like what's actually going on is that PPSEN0 modifies the behavior of bits [0:3] - when PPSEN0 is 0, bits 0:3 act as PPSCTRL, but when PPSEN0 is 1, bits 0:3 act as PPSCMD.

That being said, I have no idea what the right course for handling this in the crate is.

ehntoo replied:
Looks like some Intel/Altera parts use the same (or similar) Synopsys IP for their ethernet peripherals, and their docs are a little clearer on this point: https://www.intel.co.jp/content/dam/altera-www/global/ja_JP/pdfs/literature/hb/cyclone-v/hps.pdf

ehntoo commented:
Thanks for the thorough review! I addressed the issues you caught save for the oddity of PPSCMD/PPSCTRL. Perhaps the best way to handle that is just to change the field name to combine the two?

adamgreig commented:
Updates look good, thanks. Maybe just call it `PPSCMD` and document that it's an alias for `PPSCTRL`?

ehntoo commented:
Updated the MACPPSCR register to remove the duplicate. I wasn't quite sure how to smash the descriptions together, but this seemed fine to me.
